[BUG] Partition columns mishandled when Parquet read is coalesced and chunked

**Describe the bug**
When the coalesce reader concatenates input splits across partitions, it must construct the partition columns in a non-trivial way, tracking how many rows correspond to each partition value.  This works fine when the batch comes back as a single batch, but when the read consists of multiple batches the logic doesn't take that into account.  The result is an attempt to concatenate the generated partition columns with as many rows as the full table to a batch that has columns with fewer rows.  If the resulting, invalid batch subsequently is marshaled into a cudf Table, one of two errors will result.  If Java assertions are enabled, it will fail with:
```
java.lang.AssertionError: All columns should have the same number of rows
```
if assertions are not enabled, it will fail in libcudf with:
```
cpp/src/table/table_view.cpp:35: Column size mismatch.
```

**Steps/Code to reproduce bug**
Modify the `test_parquet_simple_partitioned_read` test to add the following config to the chunked reader config setup:
```
'spark.rapids.sql.batchSizeBytes': 100
```

Hi @jlowe, can you share more detail about the repro case ?

I tried to add this config, but some tests failed due to different output of CPU and GPU, not the error about column size mismatch.
```diff
diff --git a/integration_tests/src/main/python/parquet_test.py b/integration_tests/src/main/python/parquet_test.py
index efad1075a..2617f7bc3 100644
--- a/integration_tests/src/main/python/parquet_test.py
+++ b/integration_tests/src/main/python/parquet_test.py
@@ -419,7 +419,8 @@ def test_parquet_simple_partitioned_read(spark_tmp_path, v1_enabled_list, reader
             lambda spark : gen_df(spark, gen_list).write.parquet(third_data_path),
             conf=rebase_write_corrected_conf)
     data_path = spark_tmp_path + '/PARQUET_DATA'
-    all_confs = copy_and_update(reader_confs, {'spark.sql.sources.useV1SourceList': v1_enabled_list})
+    all_confs = copy_and_update(reader_confs, {'spark.sql.sources.useV1SourceList': v1_enabled_list,
+                                               'spark.rapids.sql.batchSizeBytes': 100})
     assert_gpu_and_cpu_are_equal_collect(
             lambda spark : spark.read.parquet(data_path),
             conf=all_confs)
``` 

```bash
        elif (t is dict):
            # The order of key/values is not guaranteed in python dicts, nor are they guaranteed by Spark
            # so sort the items to do our best with ignoring the order of dicts
            cpu_items = list(cpu.items()).sort(key=_RowCmp)
            gpu_items = list(gpu.items()).sort(key=_RowCmp)
            _assert_equal(cpu_items, gpu_items, float_check, path + ["map"])
        elif (t is int):
>           assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)
E           AssertionError: GPU and CPU int values are different at [4096, 'key']

../../src/main/python/asserts.py:77: AssertionError
...
=========================== short test summary info ============================
FAILED ../../src/main/python/parquet_test.py::test_parquet_simple_partitioned_read[-reader_confs2]
FAILED ../../src/main/python/parquet_test.py::test_parquet_simple_partitioned_read[-reader_confs3]
FAILED ../../src/main/python/parquet_test.py::test_parquet_simple_partitioned_read[-reader_confs4]
FAILED ../../src/main/python/parquet_test.py::test_parquet_simple_partitioned_read[-reader_confs7]
FAILED ../../src/main/python/parquet_test.py::test_parquet_simple_partitioned_read[parquet-reader_confs2]
FAILED ../../src/main/python/parquet_test.py::test_parquet_simple_partitioned_read[parquet-reader_confs3]
FAILED ../../src/main/python/parquet_test.py::test_parquet_simple_partitioned_read[parquet-reader_confs4]
FAILED ../../src/main/python/parquet_test.py::test_parquet_simple_partitioned_read[parquet-reader_confs7]
```

--- a/integration_tests/src/main/python/parquet_test.py
+++ b/integration_tests/src/main/python/parquet_test.py
@@ -399,7 +399,8 @@ def test_parquet_read_round_trip_legacy(spark_tmp_path, parquet_gens, v1_enabled
 
 @pytest.mark.parametrize('reader_confs', reader_opt_confs)
 @pytest.mark.parametrize('v1_enabled_list', ["", "parquet"])
-def test_parquet_simple_partitioned_read(spark_tmp_path, v1_enabled_list, reader_confs):
+@pytest.mark.parametrize('batch_size', [100, INT_MAX])
+def test_parquet_simple_partitioned_read(spark_tmp_path, v1_enabled_list, reader_confs, batch_size):
     # Once https://github.com/NVIDIA/spark-rapids/issues/133 and https://github.com/NVIDIA/spark-rapids/issues/132 are fixed
     # we should go with a more standard set of generators
     parquet_gens = [byte_gen, short_gen, int_gen, long_gen, float_gen, double_gen,
@@ -419,7 +420,9 @@ def test_parquet_simple_partitioned_read(spark_tmp_path, v1_enabled_list, reader
             lambda spark : gen_df(spark, gen_list).write.parquet(third_data_path),
             conf=rebase_write_corrected_conf)
     data_path = spark_tmp_path + '/PARQUET_DATA'
-    all_confs = copy_and_update(reader_confs, {'spark.sql.sources.useV1SourceList': v1_enabled_list})
+    all_confs = copy_and_update(reader_confs,
+            {'spark.sql.sources.useV1SourceList': v1_enabled_list,
+             'spark.rapids.sql.batchSizeBytes': batch_size})
     assert_gpu_and_cpu_are_equal_collect(
             lambda spark : spark.read.parquet(data_path),
             conf=all_confs)

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuColumnarBatchIterator.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuColumnarBatchIterator.scala
@@ -17,6 +17,8 @@
 package com.nvidia.spark.rapids
 
 import org.apache.spark.TaskContext
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.types.StructType
 import org.apache.spark.sql.vectorized.ColumnarBatch
 
 /**
@@ -74,6 +76,85 @@ class SingleGpuColumnarBatchIterator(private var batch: ColumnarBatch)
     if (batch != null) {
       batch.close()
       batch = null
+    }
+  }
+}
+
+/**
+ * An iterator that appends partition columns to each batch in the input iterator.
+ *
+ * This iterator will correctly handle multiple partition values for each partition column
+ * for a chunked read.
+ *
+ * @param inputIter the input iterator of GPU columnar batch
+ * @param partValues partition values collected from all the batches in the input iterator
+ * @param partRowNums row numbers collected from all the batches in the input iterator, it
+ *                    should have the same size with "partValues".
+ * @param partSchema the partition schema
+ */
+class GpuColumnarBatchWithPartitionValuesIterator(
+    inputIter: GpuColumnarBatchIterator,
+    partValues: Array[InternalRow],
+    partRowNums: Array[Long],
+    partSchema: StructType) extends Iterator[ColumnarBatch] with Arm {
+  assert(partValues.length == partRowNums.length)
+
+  private var leftValues: Array[InternalRow] = partValues
+  private var leftRowNums: Array[Long] = partRowNums
+
+  override def hasNext: Boolean = inputIter.hasNext
+
+  override def next(): ColumnarBatch = {
+    if (!hasNext) throw new NoSuchElementException()
+    val batch = inputIter.next()
+    val (readPartValues, readPartRows) = closeOnExcept(batch) { _ =>
+      computeValuesAndRowNumsForBatch(batch.numRows())
+    }
+    MultiFileReaderUtils.addMultiplePartitionValuesAndClose(batch, readPartValues,
+      readPartRows, partSchema)
+  }
+
+  private[this] def computeValuesAndRowNumsForBatch(batchRowNum: Int):
+      (Array[InternalRow], Array[Long]) = {
+    val leftTotalRowNum = leftRowNums.sum
+    if (leftTotalRowNum == batchRowNum) {
+      // case a) All is read
+      (leftValues, leftRowNums)
+    } else if (leftTotalRowNum > batchRowNum) {
+      // case b) Partial is read
+      var consumedRowNum = 0L
+      var pos = 0
+      // 1: Locate the position for the current read
+      while (consumedRowNum < batchRowNum) {
+        // Not test "pos < leftRowNums.length" here because this is ensured
+        // by "leftRowNum > readNum"
+        consumedRowNum += leftRowNums(pos)
+        pos += 1
+      }
+      // 2: Split the arrays of values and row numbers for the current read
+      val (readValues, remainValues) = leftValues.splitAt(pos)
+      val (readRowNums, remainRowNums) = leftRowNums.splitAt(pos)
+      if (consumedRowNum == batchRowNum) {
+        // Good luck! Just at the edge of a partition
+        leftValues = remainValues
+        leftRowNums = remainRowNums
+      } else { // consumedRowNum > readNum, and pos > 0
+        // A worse case, inside a partition, need to correct the splits.
+        // e.g.
+        //    Row numbers: [2, 3, 2]
+        //    Read row number: 4,
+        // The original split result is:  [2, 3] and [2]
+        // And the corrected output is: [2, 2] and [1, 2]
+        val remainRowNumForSplitPart = consumedRowNum - batchRowNum
+        leftRowNums = remainRowNumForSplitPart +: remainRowNums
+        readRowNums(pos - 1) = readRowNums(pos - 1) - remainRowNumForSplitPart
+        leftValues = readValues(pos - 1) +: remainValues
+      }
+      (readValues, readRowNums)
+    } else { // leftTotalRowNum < readNum
+      // This should not happen, so throw an exception
+      throw new IllegalStateException(s"Partition row number <$leftTotalRowNum> " +
+        s"does not match that of the read batch <$batchRowNum>.")
     }
   }
 }

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala
@@ -85,17 +85,8 @@ trait MultiFileReaderFunctions extends Arm {
       batch: ColumnarBatch,
       inPartitionValues: InternalRow,
       partitionSchema: StructType): ColumnarBatch = {
-    if (partitionSchema.nonEmpty) {
-      val partitionValues = inPartitionValues.toSeq(partitionSchema)
-      val partitionScalars = ColumnarPartitionReaderWithPartitionValues
-          .createPartitionValues(partitionValues, partitionSchema)
-      withResource(partitionScalars) { scalars =>
-        ColumnarPartitionReaderWithPartitionValues.addPartitionValues(batch, scalars,
-          GpuColumnVector.extractTypes(partitionSchema))
-      }
-    } else {
-      batch
-    }
+    MultiFileReaderUtils.addSinglePartitionValuesAndClose(batch, inPartitionValues,
+      partitionSchema)
   }
 
   @scala.annotation.nowarn(
@@ -144,7 +135,7 @@ object MultiFileReaderThreadPool extends Logging {
   }
 }
 
-object MultiFileReaderUtils {
+object MultiFileReaderUtils extends Arm {
 
   private implicit def toURI(path: String): URI = {
     try {
@@ -174,6 +165,83 @@ object MultiFileReaderUtils {
       anyAlluxioPathsReplaced: Boolean = false): Boolean =
   !coalescingEnabled || (multiThreadEnabled &&
     (!anyAlluxioPathsReplaced && hasPathInCloud(files, cloudSchemes)))
+
+  /**
+   * Add all partition values found to the batch. There could be more then one partition
+   * value in the batch so we have to build up columns with the correct number of rows
+   * for each partition value.
+   *
+   * @param batch             - input batch, will be closed after the call returns
+   * @param partitionValues   - partition values collected from the batch
+   * @param partitionRows     - row numbers collected from the batch, and it should have
+   *                            the same size with "partitionValues"
+   * @param partitionSchema   - the partition schema
+   * @return a new columnar batch with partition values
+   */
+  def addMultiplePartitionValuesAndClose(batch: ColumnarBatch,
+      partitionValues: Array[InternalRow],
+      partitionRows: Array[Long],
+      partitionSchema: StructType): ColumnarBatch = {
+    if (partitionSchema.nonEmpty) {
+      if (partitionValues.length > 1) {
+        concatAndAddPartitionColsToBatchAndClose(batch, partitionRows, partitionValues,
+          partitionSchema)
+      } else {
+        // single partition, add like other readers
+        addSinglePartitionValuesAndClose(batch, partitionValues.head, partitionSchema)
+      }
+    } else {
+      batch
+    }
+  }
+
+  def addSinglePartitionValuesAndClose(batch: ColumnarBatch, partitionValues: InternalRow,
+      partitionSchema: StructType): ColumnarBatch = {
+    if (partitionSchema.nonEmpty) {
+      val partitionScalars = closeOnExcept(batch) { _ =>
+        ColumnarPartitionReaderWithPartitionValues.createPartitionValues(
+          partitionValues.toSeq(partitionSchema), partitionSchema)
+      }
+      withResource(partitionScalars) { _ =>
+        ColumnarPartitionReaderWithPartitionValues.addPartitionValues(batch, partitionScalars,
+          GpuColumnVector.extractTypes(partitionSchema))
+      }
+    } else {
+      batch
+    }
+  }
+
+  private def concatAndAddPartitionColsToBatchAndClose(cb: ColumnarBatch, partRows: Array[Long],
+      partValues: Array[InternalRow], partSchema: StructType): ColumnarBatch = {
+    withResource(cb) { _ =>
+      closeOnExcept(buildPartitionsColumns(partRows, partValues, partSchema)) { partsCols =>
+        ColumnarPartitionReaderWithPartitionValues.addGpuColumVectorsToBatch(cb, partsCols)
+      }
+    }
+  }
+
+  private def buildPartitionsColumns(partRows: Array[Long], partValues: Array[InternalRow],
+      partSchema: StructType): Array[GpuColumnVector] = {
+    // build the partitions vectors for all partitions within each column
+    // and concatenate those together then go to the next column
+    val rowNumsAndValues = partRows.zip(partValues)
+    closeOnExcept(new Array[GpuColumnVector](partSchema.length)) { partsCols =>
+      for ((field, colIndex) <- partSchema.zipWithIndex) {
+        val dataType = field.dataType
+        withResource(new Array[ColumnVector](partValues.length)) { onePartCols =>
+          rowNumsAndValues.zipWithIndex.foreach { case ((rowNum, valueRow), partId) =>
+            val singleValue = valueRow.get(colIndex, dataType)
+            withResource(GpuScalar.from(singleValue, dataType)) { oneScalar =>
+              onePartCols(partId) = ColumnVector.fromScalar(oneScalar, rowNum.toInt)
+            }
+          }
+          partsCols(colIndex) = GpuColumnVector.from(
+            ColumnVector.concatenate(onePartCols: _*), field.dataType)
+        }
+      }
+      partsCols
+    }
+  }
 }
 
 /**
@@ -871,11 +939,9 @@ abstract class MultiFileCoalescingPartitionReaderBase(
           currentChunkMeta.readSchema, currentChunkMeta.extraInfo)
         CachedGpuBatchIterator(tableReader, colTypes, spillCallback)
       }
-      batchIter.map { cb =>
-        // we have to add partition values here for this batch, we already verified that
-        // its not different for all the blocks in this batch
-        withResource(addAllPartitionValues(cb, currentChunkMeta.allPartValues,
-          currentChunkMeta.rowsPerPartition, partitionSchema)) { withParts =>
+      new GpuColumnarBatchWithPartitionValuesIterator(batchIter, currentChunkMeta.allPartValues,
+          currentChunkMeta.rowsPerPartition, partitionSchema).map { withParts =>
+        withResource(withParts) { _ =>
           finalizeOutputBatch(withParts, currentChunkMeta.extraInfo)
         }
       }
@@ -1092,76 +1158,4 @@ abstract class MultiFileCoalescingPartitionReaderBase(
     CurrentChunkMeta(currentClippedSchema, currentReadSchema, currentChunk,
       numRows, rowsPerPartition.toArray, allPartValues.toArray, extraInfo)
   }
-
-  /**
-   * Add all partition values found to the batch. There could be more then one partition
-   * value in the batch so we have to build up columns with the correct number of rows
-   * for each partition value.
-   *
-   * @param batch - columnar batch to append partition values to
-   * @param inPartitionValues - array of partition values
-   * @param rowsPerPartition - the number of rows that require each partition value
-   * @param partitionSchema - schema of the partitions
-   * @return
-   */
-  private def addAllPartitionValues(
-      batch: ColumnarBatch,
-      inPartitionValues: Array[InternalRow],
-      rowsPerPartition: Array[Long],
-      partitionSchema: StructType): ColumnarBatch = {
-    assert(rowsPerPartition.length == inPartitionValues.length)
-    if (partitionSchema.nonEmpty) {
-      val numPartitions = inPartitionValues.length
-      if (numPartitions > 1) {
-        concatAndAddPartitionColsToBatch(batch, rowsPerPartition, inPartitionValues)
-      } else {
-        // single partition, add like other readers
-        addPartitionValues(batch, inPartitionValues.head, partitionSchema)
-      }
-    } else {
-      batch
-    }
-  }
-
-  private def concatAndAddPartitionColsToBatch(
-      cb: ColumnarBatch,
-      rowsPerPartition: Array[Long],
-      inPartitionValues: Array[InternalRow]): ColumnarBatch = {
-    withResource(cb) { _ =>
-      closeOnExcept(buildAndConcatPartitionColumns(rowsPerPartition, inPartitionValues)) {
-        allPartCols =>
-          ColumnarPartitionReaderWithPartitionValues.addGpuColumVectorsToBatch(cb, allPartCols)
-      }
-    }
-  }
-
-  private def buildAndConcatPartitionColumns(
-      rowsPerPartition: Array[Long],
-      inPartitionValues: Array[InternalRow]): Array[GpuColumnVector] = {
-    val numCols = partitionSchema.fields.length
-    val allPartCols = new Array[GpuColumnVector](numCols)
-    // build the partitions vectors for all partitions within each column
-    // and concatenate those together then go to the next column
-    for ((field, colIndex) <- partitionSchema.fields.zipWithIndex) {
-      val dataType = field.dataType
-      withResource(new Array[GpuColumnVector](inPartitionValues.length)) {
-        partitionColumns =>
-          for ((rowsInPart, partIndex) <- rowsPerPartition.zipWithIndex) {
-            val partInternalRow = inPartitionValues(partIndex)
-            val partValueForCol = partInternalRow.get(colIndex, dataType)
-            val partitionScalar = GpuScalar.from(partValueForCol, dataType)
-            withResource(partitionScalar) { scalar =>
-              partitionColumns(partIndex) = GpuColumnVector.from(
-                ai.rapids.cudf.ColumnVector.fromScalar(scalar, rowsInPart.toInt),
-                dataType)
-            }
-          }
-          val baseOfCols = partitionColumns.map(_.getBase)
-          allPartCols(colIndex) = GpuColumnVector.from(
-            ColumnVector.concatenate(baseOfCols: _*), field.dataType)
-      }
-    }
-    allPartCols
-  }
-
 }

